### PR TITLE
Add portfolio method notebook

### PR DESCRIPTION
**Pull request by AlexGoodallAI**

### CODE REVIEW

File: trieste/acquisition/function/function.py [hunk 1]
```diff
@@ -87,7 +87,7 @@ def update_acquisition_function(
         tf.debugging.assert_positive(len(dataset), message="Dataset must be populated.")
         mean, _ = model.predict(dataset.query_points)
         eta = tf.reduce_min(mean, axis=0)[0]
-        function = probability_below_threshold(model, eta)
+        function.update(eta)  # type: ignore
```
Comment (uri-granta) [inline trieste/acquisition/function/function.py:91]: (can you also add a `tf.debugging.Assert(isinstance(function, probability_below_threshold), [tf.constant([])])`?)

File: docs/notebooks/qhsri-tutorial.pct.py [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+#     display_name: Python 3 (ipykernel)
+#     language: python
+#     name: python3
+# ---
+
+# %% [markdown]
+# # Batch HSRI Tutorial
+
+# %% [markdown]
+# Batch Hypervolume Sharpe Ratio Indicator (qHSRI) is a method for picking a batch of query points during Bayesian Optimisation. It makes use of the Sharpe Ratio, a portfolio selection method used in investment selection to carefully balance risk and reward.
```
Comment (vpicheny) [inline docs/notebooks/qhsri-tutorial.pct.py]: we should cite the reference paper upfront, 
Batch Hypervolume Sharpe Ratio Indicator (qHSRI) is a method **proposed by ....** for picking

File: docs/notebooks/qhsri-tutorial.pct.py [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+
+# %%
+from trieste.models.gpflow import GaussianProcessRegression, build_gpr
+
+# Set up model
+gpflow_model = build_gpr(initial_data, search_space, likelihood_variance=1e-7)
+model = GaussianProcessRegression(gpflow_model)
+
+# %% [markdown]
+# Now consider how we might want to select a batch of $q$ query points to observe. It would be useful for some of these to be "safe bets" that we think are very likely to provide an improvement. It would also be valuable for some of these to be sampling parts of the space that we have no idea what the expected observed value is (i.e. high variance). This problem is very similar to that encountered in building finance portfolios, where you want a mix of high  risk/high reward and low risk/low reward assets. You would also want to know how much of your total capital to invest in each asset.
```
Comment (vpicheny) [inline docs/notebooks/qhsri-tutorial.pct.py]: perhaps weird spacing in **high risk/high reward and low risk/low reward assets**

Comment (vpicheny) [inline docs/notebooks/qhsri-tutorial.pct.py]: "that we think are very likely to provide an improvement."

that we think are very likely to provide good values (i.e. low predicted mean)

File: docs/notebooks/qhsri-tutorial.pct.py [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+#
+# We can sample from the model to determine the mean and standard deviation of points. For example if we randomly sample 1000 points from the space, we can then plot the mean and standard deviation. As we actually want to maximise standard deviation, we will multiply by -1 once sampled
```
Comment (vpicheny) [inline docs/notebooks/qhsri-tutorial.pct.py]: "We can sample from the model"

that's not exactly correct. Why not:

To visualise the different trade-offs, we can sample from the input space, compute the predictive mean and variance at those locations, and plot the mean against minus the standard deviation (so that both quantities need to be minimised)

File: docs/notebooks/qhsri-tutorial.pct.py [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+
+# %%
+uniform_points = tf.random.uniform([1000, 2], dtype="float64")
```
Comment (vpicheny) [inline docs/notebooks/qhsri-tutorial.pct.py]: might be more triesty to use search_space.sample(1000)

File: docs/notebooks/qhsri-tutorial.pct.py [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+plt.xlabel("Mean")
+plt.ylabel("Negative std")
+plt.show()
+
+print(f"There are {len(uniform_non_dominated)} non-dominated points")
+
+# %% [markdown]
+# We can see that there's only a few non-dominated points to choose from the select the next batch. This set of non-dominated points is the Pareto front in the optimisation task of minimising mean and maximising standard deviation.
+#
+# This means we can improve on this by using optimisation rather than random sampling to pick the points that we will select our batch from. In trieste we use the NSGA-II multi-objective optimisation method from the [pymoo](https://pymoo.org/) library to find the Pareto front.
```
Comment (uri-granta) [inline docs/notebooks/qhsri-tutorial.pct.py:102]: Mention (either here or elsewhere) how to install the optional qhsri dependency?

**Review by uri-granta — APPROVED** [review]
LGTM!

File: docs/notebooks/qhsri-tutorial.pct.py [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+plt.close()
+
+plt.scatter(uniform_pts_mean, uniform_pts_std)
+plt.title("Mean vs negative std of uniformly sampled points")
+plt.xlabel("Mean")
+plt.ylabel("Negative std")
+plt.show()
+
+# %% [markdown]
+# Since we only want the best points in terms of the risk reward tradeoff, we want points that have low means and high standard deviations. So we remove any points that are dominated by another point. A point `a` is dominated by another point `b` if `b.mean` <= `a.mean` and `b.std` >= `a.std`.
```
Comment (vpicheny) [inline docs/notebooks/qhsri-tutorial.pct.py]: Since we only want the best points in terms of the risk-reward trade-off, we can remove all the points that are not optimal in the Pareto sense, i.e. the points that are dominated by another point. A point `a` is dominated by another point `b` if `b.mean` <= `a.mean` and `b.std` >= `a.std`.

File: docs/notebooks/qhsri-tutorial.pct.py [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+plt.show()
+
+print(f"There are {len(uniform_non_dominated)} non-dominated points")
+
+# %% [markdown]
+# We can see that there's only a few non-dominated points to choose from the select the next batch. This set of non-dominated points is the Pareto front in the optimisation task of minimising mean and maximising standard deviation.
+#
+# This means we can improve on this by using optimisation rather than random sampling to pick the points that we will select our batch from. In trieste we use the NSGA-II multi-objective optimisation method from the [pymoo](https://pymoo.org/) library to find the Pareto front.
+#
+# The `BatchHypervolumeSharpeRatioIndicator` acqusition rule makes use of the `_MeanStdTradeoff` class, which expresses the optimisation problem in the pymoo framework. Pymoo is then used to run the optimisation.
```
Comment (vpicheny) [inline docs/notebooks/qhsri-tutorial.pct.py]: acquisition

File: docs/notebooks/qhsri-tutorial.pct.py [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
+plt.scatter(result.X[:, 0], result.X[:, 1])
+plt.title("Points optimised for mean and std")
+plt.show()
+
+plt.scatter(result.F[:, 0], result.F[:, 1])
+plt.title("Mean vs std of optimised points")
+plt.show()
+
+# %% [markdown]
+# We can check the non-dominated points again, and see that this is a great improvement.
```
Comment (vpicheny) [inline docs/notebooks/qhsri-tutorial.pct.py]: ... , and see that the outcome of NSGA-II is much better than the randomly sampled ones.

(assuming this is the case)

File: docs/notebooks/qhsri-tutorial.pct.py [hunk 10]
```diff
@@ -0,0 +1,232 @@
...
+# We can see that there's only a few non-dominated points to choose from the select the next batch. This set of non-dominated points is the Pareto front in the optimisation task of minimising mean and maximising standard deviation.
+#
+# This means we can improve on this by using optimisation rather than random sampling to pick the points that we will select our batch from. In trieste we use the NSGA-II multi-objective optimisation method from the [pymoo](https://pymoo.org/) library to find the Pareto front.
+#
+# The `BatchHypervolumeSharpeRatioIndicator` acqusition rule makes use of the `_MeanStdTradeoff` class, which expresses the optimisation problem in the pymoo framework. Pymoo is then used to run the optimisation.
+#
+# NSGA-II is a genetic algorithm, and so we need to define a population size and number of generations.
+
+# %%
+import numpy as np
```
Comment (vpicheny) [inline docs/notebooks/qhsri-tutorial.pct.py:108]: could we replace all of this by just calling the new rule?

**Review by vpicheny — APPROVED** [review]
looks good!
